pkg/provider/azure_utils.go

```diff
@@ -448,50 +448,81 @@ func getResourceByIPFamily(resource string, isDualStack, isIPv6 bool) string {
 }
 
 // isFIPIPv6 checks if the frontend IP configuration is of IPv6.
-func (az *Cloud) isFIPIPv6(service *v1.Service, fip *network.FrontendIPConfiguration, isInternal bool) (bool, error) {
-	if isInternal {
-		if fip.FrontendIPConfigurationPropertiesFormat != nil {
-			if fip.FrontendIPConfigurationPropertiesFormat.PrivateIPAddressVersion != "" {
-				return fip.FrontendIPConfigurationPropertiesFormat.PrivateIPAddressVersion == network.IPv6, nil
-			}
-			if fip.FrontendIPConfigurationPropertiesFormat.PrivateIPAddress != nil {
-				return net.ParseIP(*fip.FrontendIPConfigurationPropertiesFormat.PrivateIPAddress).To4() == nil, nil
-			}
+func (az *Cloud) isFIPIPv6(service *v1.Service, pipRG string, fip *network.FrontendIPConfiguration) (bool, error) {
+	isServiceOwnsFrontendIP, isPrimaryService, err := az.serviceOwnsFrontendIP(*fip, service)
+	if err != nil {
+		return false, err
+	}
+	if isServiceOwnsFrontendIP && isPrimaryService {
+		isDualStack := isServiceDualStack(service)
+		if isDualStack {
+			return managedResourceHasIPv6Suffix(pointer.StringDeref(fip.Name, "")), nil
 		}
-		klog.Errorf("Checking IP Family of frontend IP configuration %q of internal Service but "+
-			"it is not clear. It's considered to be IPv4",
-			pointer.StringDeref(fip.Name, ""))
-		return false, nil
+		if len(service.Spec.IPFamilies) != 1 {
+			return false, fmt.Errorf("number of Service %q IPFamilies should be one", service.Name)
+		}
+		return service.Spec.IPFamilies[0] == v1.IPv6Protocol, nil
 	}
-	var fipPIPID string
-	if fip.FrontendIPConfigurationPropertiesFormat != nil && fip.FrontendIPConfigurationPropertiesFormat.PublicIPAddress != nil {
-		fipPIPID = pointer.StringDeref(fip.FrontendIPConfigurationPropertiesFormat.PublicIPAddress.ID, "")
+
+	// First check private
+	// For an external LB, the FIP's private IP address is empty. The logic will
+	// go through this part and check public IP address then.
+	if fip.FrontendIPConfigurationPropertiesFormat != nil {
```

Comment:
I checked all the refs of this function. If we can make sure the fip is always owned by the service, the pip check is not needed.

Reply:
The FIP also needs to be `isPrimaryService==true`. If it is the secondary Service, I think the pip check is still needed.

Reply:
If the fip is owned by a single stack service, we can check the service instead of pip, no matter if it is primary or secondary.

Reply:
It makes sense for single stack svc. Updated.